subsys/nfc/lib/platform.c

```diff
@@ -64,46 +65,55 @@ nrfx_err_t nfc_platform_setup(void)
 }
 
 
-static nrfx_err_t nfc_platform_tagheader_get(uint8_t index,
-					     uint32_t *tag_header)
+static nrfx_err_t nfc_platform_tagheaders_get(uint32_t tag_header[3])
 {
-	const uint32_t tag_header_addresses[] = {
-#ifdef CONFIG_TRUSTED_EXECUTION_NONSECURE
-		(uint32_t) &NRF_FICR_S->NFC.TAGHEADER0,
-		(uint32_t) &NRF_FICR_S->NFC.TAGHEADER1,
-		(uint32_t) &NRF_FICR_S->NFC.TAGHEADER2
+	__IOM FICR_NFC_Type *ficr_nfc =
+#if defined(NRF_TRUSTZONE_NONSECURE)
+				&NRF_FICR_S->NFC;
 #else
-		(uint32_t) &NRF_FICR->NFC.TAGHEADER0,
-		(uint32_t) &NRF_FICR->NFC.TAGHEADER1,
-		(uint32_t) &NRF_FICR->NFC.TAGHEADER2
+				&NRF_FICR->NFC;
 #endif
-	};
 
-	if (index >= ARRAY_SIZE(tag_header_addresses)) {
-		LOG_ERR("Tag Header index out of bounds");
-		return NRFX_ERROR_INVALID_ADDR;
-	}
 
 #ifdef CONFIG_TRUSTED_EXECUTION_NONSECURE
+#if defined(CONFIG_BUILD_WITH_TFM)
+	uint32_t err = 0;
+	enum tfm_platform_err_t plt_err;
+	FICR_NFC_Type ficr_nfc_ns;
+
+	plt_err = tfm_platform_mem_read(&ficr_nfc_ns, (uint32_t)ficr_nfc,
+					sizeof(ficr_nfc_ns), &err);
+	if (plt_err != TFM_PLATFORM_ERR_SUCCESS || err != 0) {
+		LOG_ERR("Could not read FICR NFC Tag Header (plt_err %d, err: %d)",
+			plt_err, err);
+		return NRFX_ERROR_INTERNAL;
+	}
+
+	ficr_nfc = &ficr_nfc_ns;
+#elif defined(CONFIG_SPM_SERVICE_READ)
 	int err;
+	FICR_NFC_Type ficr_nfc_ns;
 
-	// TODO: Read all three headers at once to avoid the 300us IPC
-	// overhead and add support for the TF-M equivalent of
-	// spm_request_read
-	err = spm_request_read(tag_header, tag_header_addresses[index],
-			       sizeof(uint32_t));
+	err = spm_request_read(&ficr_nfc_ns, (uint32_t)ficr_nfc,
+			       sizeof(ficr_nfc_ns));
 	if (err != 0) {
-		LOG_ERR("Could not read NFC Tag Header FICR (err: %d)", err);
-		return NRFX_ERROR_INVALID_ADDR;
+		LOG_ERR("Could not read FICR NFC Tag Header (err: %d)", err);
+		return NRFX_ERROR_INTERNAL;
 	}
+
+	ficr_nfc = &ficr_nfc_ns;
 #else
-	*tag_header = *((uint32_t *) tag_header_addresses[index]);
+	return NRFX_ERROR_FORBIDDEN;
```

Comment:
I think a build error would be more appropriate.

#error "Unexpected configuration"

could put it in the ifdef we do on #includes to not disturb the runtime code. And omit the else here.

Reply:
I didn't think it would be possible to use anything here if tag headers are unavailable.

But perhaps a domain expert can check this assumption.

Essentially, if it is necessary to gracefully handle this condition at runtime then we need it, otherwise it's easier
for a user to debug a build error than a runtime error.

Reply:
Changed to build error.